### ISSUE TITLE
Output of template module inconsistent return vars between runs

### ISSUE THREAD

**Issue by Rob-Johnson** [ansible/ansible#10190]
Hi,

Issue Type: "Bug Report"

Ansible Version: ansible 1.9 (devel 68d45122c2)

Environment:
Host: Tested on both Ubuntu 14.04 and OSX 10.10 Yosemite
Remote: Not applicable

Summary:

I'm using the template module to write files locally and registering the created files. I'm then using the `path` field of those registered variables in another template. When accessing the registered variable on the first run, it does not contain the `path` variable. When the task is re-run, it works fine.

Steps To Reproduce:

Here is a gist https://gist.github.com/Rob-Johnson/8408ac61853671f2f95b to reproduce. It is in the rendering of `dockerfile.j2` that the error occurs, particularly when accessing the `file.path` variable.

Expected Results:

The registered variable contains the expected fields.

On the first run, this fails with the error:

Actual Results:

```
 `fatal: [localhost -> 127.0.0.1] => {'msg': "AnsibleUndefinedVariable: One or more undefined variables: 'dict object' has no attribute 'path'", 'failed': True}`
```

It is successful on the second run.

Thanks!

EDIT:

Here's the output of adding `-debug: var=files` to the `site.yml` playbook:

First Run:

``` json

ok: [localhost] => {
    "var": {
        "files": {
            "changed": true,
            "msg": "All items completed",
            "results": [
                {
                    "changed": true,
                    "checksum": "632a23bfb56e421ba2e58793ba6d6004c4f9cdbe",
                    "dest": "/tmp/tmp/logspout-config/logspout.syslog_servers",
                    "gid": 20,
                    "group": "staff",
                    "invocation": {
                        "module_args": "src=routing.j2 dest=\"/tmp/tmp/logspout-config/logspout.syslog_servers\"",
                        "module_name": "template"
                    },
                    "item": [
                        0,
                        "logspout.syslog_servers"
                    ],
                    "md5sum": "0ef436e907e00c46fcd937abdf358fdd",
                    "mode": "0644",
                    "owner": "rob",
                    "size": 99,
                    "src": "/Users/rob/.ansible/tmp/ansible-tmp-1423495808.83-168578190674517/source",
                    "state": "file",
                    "uid": 501
                }
            ]
        }
    }
}
```

Second Run:

``` json

ok: [localhost] => {
    "var": {
        "files": {
            "changed": false,
            "msg": "All items completed",
            "results": [
                {
                    "changed": false,
                    "gid": 20,
                    "group": "staff",
                    "invocation": {
                        "module_args": "src=routing.j2 dest=\"/tmp/tmp/logspout-config/logspout.syslog_servers\"",
                        "module_name": "template"
                    },
                    "item": [
                        0,
                        "logspout.syslog_servers"
                    ],
                    "mode": "0644",
                    "owner": "rob",
                    "path": "/tmp/tmp/logspout-config/logspout.syslog_servers",
                    "size": 99,
                    "state": "file",
                    "uid": 501
                }
            ]
        }
    }
}
```

**Comment by jamiemccrindle** [comment]
+1

**Comment by abadger** [comment]
First run has:
- src
- dest
- md5sum
- checksum

Second run has:
- path

@bcoca Seems best to normalize the names of these values but I'm not sure if we need to keep the old values for backwards compat?  Do we want to normalize them at their source (the file and the copy modules) or just within template (with something like: if 'dest' in res and 'path' not in res: res['path'] = res['dest'] ).

**Comment by bcoca** [comment]
the issue is that template is not the only module returning data, file and copy modules can also be involved here

**Comment by jimi-c** [comment]
This now plain-out fails on the second task every time in the current devel, as the `path` value is never present. It would appear that using `dest` would be correct value instead?

**Comment by neh** [comment]
When I encountered this issue recently, it appeared to be (as @bcoca said above) that different modules would return data with different value names. In my case, when the template module changed the file, the path of the file affected would be in 'dest', but if the file was not changed, it would be in 'path' (like the file module).

**Comment by bcoca** [comment]
this is still a bug, we need to normalize the output, sadly that will mean always returning path and dest so as not to break existing plays

**Comment by jlaska** [comment]
Quick workaround for my usage ...

``` yaml
- file:
    path: '{{ template_result.path|default(template_result.dest) }}'
    state: absent
  when: template_result is defined
```

**Comment by jimi-c** [comment]
This has been reported as fixed in devel/2.0, so per our normal policy we'll go ahead and close this now.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
